front/pages/api/data_sources/[user]/[name]/search.ts

```diff
@@ -0,0 +1,230 @@
+import { NextApiRequest, NextApiResponse } from "next";
+import { User, DataSource, Provider, Key } from "@app/lib/models";
+import { getServerSession } from "next-auth/next";
+import { authOptions } from "@app/pages/api/auth/[...nextauth]";
+import { Op } from "sequelize";
+import { credentialsFromProviders } from "@app/lib/providers";
+import { parse_payload, RequestParseError } from "@app/lib/http_utils";
+import { JSONSchemaType } from "ajv";
+import { Document } from "@app/lib/types";
+import { Result, Ok, Err } from "@app/lib/result";
+import { APIErrorWithStatusCode, APIError } from "@app/lib/api/error";
+
+type TagsFilter = {
+  in?: string[];
+  not?: string[];
+};
+
+type TimestampFilter = {
+  gt?: number;
+  lt?: number;
+};
+
+type SearchFilter = {
+  tags?: TagsFilter;
+  timestamp?: TimestampFilter;
+};
+
+export type DatasourceSearchQuery = {
+  query: string;
+  top_k: number;
+  full_text: boolean;
+  timestamp_gt?: number;
+  timestamp_lt?: number;
+  tags_in?: string[];
+  tags_not_in?: string[];
+};
+
+const search_query_schema: JSONSchemaType<DatasourceSearchQuery> = {
+  type: "object",
+  properties: {
+    query: { type: "string" },
+    top_k: { type: "number" },
+    full_text: { type: "boolean" },
+    timestamp_gt: { type: "number", nullable: true },
+    timestamp_lt: { type: "number", nullable: true },
+    tags_in: { type: "array", items: { type: "string" }, nullable: true },
+    tags_not_in: { type: "array", items: { type: "string" }, nullable: true },
+  },
+  required: ["query", "top_k", "full_text"],
+};
+
+const { DUST_API } = process.env;
+
+export type DatasourceSearchResponseBody = {
+  documents: Array<Document>;
+};
+
+export async function performSearch(
+  auth_user: User,
+  request_uri_user: User,
+  datasource_id: string,
+  request_payload: any
+): Promise<Result<DatasourceSearchResponseBody, APIErrorWithStatusCode>> {
```

Comment:
Didn't we say we would move that to `/lib` ?

Not great that it's typed `APIErrorWithStatusCode` but used for the internal API (it makes the internal API depend on extrenal API objects).... but eventually we'll probably unify the two error types?